Update `swift-syntax`, fix Bool switch warnings, and improve test organization & coverage

qizh opened the pull request:
This pull request updates dependencies, improves enum case encoding logic, enhances protocol constraints for numeric value coders, adds a new test for attribute misuse, and fixes compiler warnings. The changes focus on compatibility, correctness, improved diagnostics, and cleaner build output.

# Dependency Updates

- **Updated swift-syntax dependency**: Increased the upper bound to allow versions up to `603.0.0`, ensuring compatibility with newer releases
- **Added<sup>[**¹**](#swiftsyntaxmacroexpansion-details)</sup> `SwiftSyntaxMacroExpansion`**: Explicit dependency added to PluginCore to resolve Xcode's dependency scanner warnings
  <details>
    <summary>
      <h3> <a href="#swiftsyntaxmacroexpansion-details" aria-label="Permalink">📙¹</a> Dependency Addition Explanation & Reasoning</h3>
    </summary>
  
  ## `SwiftSyntaxMacroExpansion` Dependency Addition
  
  ### Problem Statement
  
  Xcode's dependency scanner was generating warnings about implicit dependencies on the `SwiftSyntaxMacroExpansion` module. While the package compiled successfully, the implicit dependency meant that Xcode couldn't properly track the module dependency graph.
  
  ### Root Cause
  
  The [`MetaCodable`](https://github.com/SwiftyLab/MetaCodable) package uses `MacroExpansionContext` and `BasicMacroExpansionContext` types from the `SwiftSyntaxMacroExpansion` module throughout the codebase, but this dependency was only declared implicitly through other swift-syntax modules. The **Swift Package Manager** resolved it transitively, but **Xcode**'s dependency scanner requires explicit declarations for all directly-imported modules.
  
  ### Solution
  
  ```swift
  .product(name: "SwiftSyntaxMacroExpansion", package: "swift-syntax")
  ``` 
  
  Should be added to the dependencies array for three targets:
  
  <ol type="1">
    <li>❇️ <strong><code>PluginCore</code></strong> (<a href="Package.swift#L30-L40">Package.swift:38</a>, <a href="Package%40swift-5.swift#L28-L39">Package@swift-5.swift:37</a>)</li>
    <li>☑️ <strong><code>MacroPlugin</code></strong> (already existed: <a href="Package.swift#L50">Package.swift#L50</a>)</li>
    <li>☑️ <strong><code>ProtocolGen</code></strong> (already existed: <a href="Package.swift#L64">Package.swift#L64</a>)</li>
  </ol>
  
  This makes the implicit dependency explicit, satisfying both the Swift Package Manager and Xcode's dependency scanner.
  
  ### Usage Across the Codebase
  
  #### 1. Macro Expansion in PluginCore
  - **[`Codable+Expansion.swift`](Sources/PluginCore/Attributes/Codable/Codable+Expansion.swift)**: [`MacroExpansionContext`](/Sources/PluginCore/Attributes/Codable/Codable%2BExpansion.swift#L29C26-L29C47) parameter in the main expansion method
  - **[`Decodable+Expansion.swift`](Sources/PluginCore/Attributes/Codable/Decodable/Decodable+Expansion.swift)**: Context passed to create type variable representations
  - **[`EnumVariable.swift`](Sources/PluginCore/Variables/Type/EnumVariable.swift)**: Context stored for later code generation phases
  - **[`CodingKeysMap.swift`](Sources/PluginCore/Variables/Type/Data/CodingKeysMap/CodingKeysMap.swift)**: Context used to generate `CodingKeys` enum declarations
  
  #### 2. Protocol Code Generation in [`ProtocolGen`](Sources/ProtocolGen/ProtocolGen.swift)
  - **[`Generate.swift`](Sources/ProtocolGen/Generate.swift)**: Direct import and usage of [`BasicMacroExpansionContext`](Sources/ProtocolGen/Generate.swift#L313)
  - Used to generate helper coder implementations for protocols
  - Creates context for both encoding and decoding code generation
  
  #### 3. Testing Infrastructure
  - **[`CodableTests.swift`](Tests/MetaCodableTests/CodableTests.swift)** and **[`ConformCodableTests.swift`](Tests/MetaCodableTests/ConformCodableTests.swift)**: Import module for macro expansion testing
  - Used in `assertMacroExpansion()` test helper functions
  - Required for verifying macro expansions produce correct code
  
  ### Impact
  
  - ✅ Resolves Xcode dependency scanner warnings
  - ✅ Makes dependency graph explicit and traceable
  - ✅ No functional changes to the codebase
  - ✅ Maintains compatibility with both Swift `5.9` and Swift `6.0` manifests
  - ✅ Aligns with Swift Package Manager best practices for explicit dependency declaration
  </details> 

# Code Improvements

- **Enum encoding logic**: Refactored enum case encoding in `EnumVariable` so cases without associated values are encoded as just the case name (without parentheses), improving generated code correctness
- **Bool switch warnings**: Skip generating default case for Bool type switches since both `true` and `false` cases are explicitly handled, eliminating "default will never be executed" warnings

# Protocol Enhancements

- **Sendable constraint**: Added `Sendable` constraint to generic requirements for numeric value coding strategies, improving safety for concurrent use

---

# Maintenance, Docs, Build Stability, Verification, and Backward-Compatible Refinements 

## Testing and Diagnostics

- **Attribute misuse test**: Added new test case to `ConformEncodableTests` to check for misuse when `@Codable` is used with `@ConformEncodable`, providing better diagnostics for attribute conflicts
- **Test cleanup**: Removed duplicate diagnostic expectations from `ConformCodableTests.swift` that were incorrectly preserved during merge conflict resolution

## Documentation Updates

- **Inline documentation**: Updated `decodeSwitchExpression` documentation to explain Bool type handling behavior
- **CHANGELOG**: Added entry documenting the Bool switch warning fix

## Build Configuration

- **Git ignore**: Added `*.dia` files to `.gitignore` (Swift dependency analysis files, similar to `.d` and `.o` files)
- **Package configuration**: Updated both `Package.swift` and `Package@swift-5.swift` to maintain Swift 5/6 compatibility

## Verification

- All tests pass without warnings
- No warnings in command line build (`swift build`)
- No warnings in Xcode build (`MetaCodable-Package` scheme)
- Both Swift 5 and Swift 6 compatibility maintained

## Context

The compiler warning fixes address issues that were particularly noticeable when:

- Using `@CodedAs` with `Bool` raw values in enums
- Building in Xcode with aggressive dependency scanning enabled

All changes maintain backward compatibility while providing cleaner build output for developers using MetaCodable.

---

# Additional Changes (Jan 30, 2026)

### 🐛 Bug Fix Refinement
- **Fixed partial Bool coverage in `TaggedEnumSwitcherVariable`**
  - The original Bool switch fix incorrectly skipped the default case for ALL Bool switches
  - Now correctly checks if **both** `true` and `false` values are present before skipping default case
  - Fixes unreachable code warnings when only one Bool value is specified (e.g., `@CodedAs<Bool>("isActive", ifTrue: SomeType.self)`)

### 🧪 Test Organization & Coverage

#### Coverage Improvements
- Added tests for previously uncovered code paths:
  - `HelperCoder`: `decodeIfPresent`, `encode`, `encodeIfPresent`
  - `ConditionalCoder`: `decodeIfPresent`, `encodeIfPresent`  
  - `DefaultSequenceElementCoding`: optional encoding/decoding methods
  - `DynamicCodableIdentifier`: `CodingKey` conformance methods
- Added partial Bool coverage tests in `CodedAsMixedTypesTests.swift`

#### Test Suite Organization
- Added `@Suite` declarations to all 50 test files for logical grouping
- Replaced 565 unnamed `@Test()` with descriptive `@Test("meaningful name")`
- Created `Tags.swift` with 30+ tag definitions for test filtering:
  - Macro tags: `.codable`, `.codedAt`, `.codedIn`, `.codedBy`, etc.
  - Test type tags: `.macroExpansion`, `.encoding`, `.decoding`, `.errorHandling`
  - Feature tags: `.helperCoders`, `.dynamicCoding`, `.generics`
- Reorganized tests into nested suites where appropriate
- Added `Tests/TODO/CoverageTracking.md` for tracking coverage progress

### 🔧 CI/CD
- Added `.github/workflows/swift.yml` for build and test automation
  - Uses default Xcode toolchain on `macos-latest`
  - Includes SwiftPM dependency caching
  - Runs `swift build` and `swift test --parallel --verbose`

### 📝 Other
- Added `.markdownlint.json` configuration

qizh committed:
Update `swift-syntax` version and improve type constraints

- [x] Bumped `swift-syntax` dependency to allow up to `603.0.0`.
- [x] Added `SendableMetatype` constraint to `NumberCodingStrategy` extension.
- [x] Refactored `caseEncodeExpr` in `EnumVariable` for consistent `FunctionCallExprSyntax` usage.

qizh committed:
Bring indentation back in `EnumVariable.swift`

- [x] Corrected indentation for the `caseEncodeExpr` closure to maintain consistent formatting.

qizh committed:
Update protocol constraint from `SendableMetatype` to `Sendable`

Replaces the `SendableMetatype` constraint with `Sendable` in the `ValueCodingStrategy` extension to align with updated protocol requirements or type definitions.

qizh committed:
Fix enum case encoding for cases without arguments

- [x] Refactors `caseEncodeExpr` to return only the case name for enum cases without associated values, avoiding unnecessary parentheses.
  This improves code generation for enums with simple cases.

qizh committed:
Align `misuseWithCodable` diagnostics with actual macro output

- [x] Update `ConformEncodableTests.misuseWithCodable` to expect four diagnostics.
- [x] Ensure diagnostic IDs, messages, fix-its, and line/column positions match emitted results:
  - [x] Two diagnostics for `@ConformEncodable` at line 1, column 1.
  - [x] Two diagnostics for `@Codable` at line 2, column 1.
  - [x] Mirrors the pattern already used in `misuseWithDecodable`.

qizh committed:
Merge remote-tracking branch 'refs/remotes/upstream/main'

#Conflicts:
#	Sources/PluginCore/Variables/Type/EnumVariable.swift
#	Tests/MetaCodableTests/ConformCodableTests.swift

qizh committed:
fix: remove unreachable default case warnings for Bool type switches

Skip generating default case in switch statements when the header type
is Bool, since both true and false cases are already explicitly handled.
This eliminates "default will never be executed" compiler warnings in
macro-generated code.

- Update decodeSwitchExpression to check for Bool type before adding default
- Update test expectation to match new behavior
- Add inline documentation explaining the Bool type handling

qizh committed:
fix: add SwiftSyntaxMacroExpansion dependency to PluginCore

Add explicit SwiftSyntaxMacroExpansion dependency to PluginCore target
to resolve Xcode's dependency scanner warnings. The scanner detects that
MacroExpansionContext (used extensively in PluginCore) requires this
module as a transitive dependency.

Applies to both Package.swift and Package@swift-5.swift for Swift 5/6
compatibility.

qizh committed:
chore: add *.dia to gitignore

Add Swift dependency analysis files (*.dia) to gitignore as they are
temporary build artifacts similar to .d and .o files.

qizh committed:
fix: remove duplicate diagnostic expectations in ConformCodableTests

Remove duplicate diagnostic test expectations that were incorrectly
preserved during merge conflict resolution.

qizh committed:
Merge pull request #1 from qizh/fix/warnings-cleanup

- [x] fix: resolve compiler warnings
- [x] merge: from `upstream/main`

soumyamahunt commented on Package.swift:
```diff
@@ -35,6 +35,7 @@ let package = Package(
                 .product(name: "SwiftDiagnostics", package: "swift-syntax"),
                 .product(name: "SwiftSyntaxBuilder", package: "swift-syntax"),
                 .product(name: "SwiftSyntaxMacros", package: "swift-syntax"),
+                .product(name: "SwiftSyntaxMacroExpansion", package: "swift-syntax"),
```
Any reason for this change?

qizh replied:
Yes, it was intentional and was noted in PR description:

> Added `SwiftSyntaxMacroExpansion`: Explicit dependency added to `PluginCore` to resolve Xcode's dependency scanner warnings

@soumyamahunt just in case, I've added a [detailed explanation in the PR description](https://github.com/SwiftyLab/MetaCodable/pull/150#swiftsyntaxmacroexpansion-details).

soumyamahunt commented on Sources/PluginCore/Variables/Type/EnumVariable.swift:
```diff
@@ -84,15 +84,19 @@ package struct EnumVariable: TypeVariable, DeclaredVariable {
-                    ) {}
-                }
-            return ExprSyntax(fExpr)
+            if args.isEmpty {
```
Any reason for this change? May be you can add some test cases that demonstrates what this fixes?

qizh replied:
@soumyamahunt, I've [updated the existing](https://github.com/SwiftyLab/MetaCodable/pull/150/changes#diff-5d90b74b4c80d0837593efc8ade4f30951404edbc3a8cd20da3fb0d266c59974L335-L336) [`WithAnyCodableLiteralEnum/expansion` test](https://github.com/SwiftyLab/MetaCodable/pull/150/changes#diff-5d90b74b4c80d0837593efc8ade4f30951404edbc3a8cd20da3fb0d266c59974L284) at the same time with this fix. 
The test should be self-explanatory, though let me know if it's not and a separate test is required.

qizh commented on CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@
 ### 🐛 Fixes
 
 * fixed optional types not detected with `valueCoder` strategy ([#141](https://github.com/SwiftyLab/MetaCodable/issues/141)) ([5873c3e](https://github.com/SwiftyLab/MetaCodable/commit/5873c3e33ab98e61c06304bfc2a2c93ab199d65d))
+* removed unreachable `default` case warnings for `Bool` type switches in macro-generated code
```
@soumyamahunt 👌 no problem
```suggestion
```

qizh commented on Package@swift-5.swift:
```diff
@@ -34,6 +34,7 @@ let package = Package(
                 .product(name: "SwiftDiagnostics", package: "swift-syntax"),
                 .product(name: "SwiftSyntaxBuilder", package: "swift-syntax"),
                 .product(name: "SwiftSyntaxMacros", package: "swift-syntax"),
+                .product(name: "SwiftSyntaxMacroExpansion", package: "swift-syntax"),
```
@soumyamahunt, it's same as [here](https://github.com/SwiftyLab/MetaCodable/pull/150/changes#r2713221190). 
Check detailed explanation and reasoning in the [just added section](https://github.com/SwiftyLab/MetaCodable/pull/150#swiftsyntaxmacroexpansion-details) in #150 description.

qizh commented on Sources/PluginCore/Variables/Enum/Switcher/TaggedEnumSwitcherVariable.swift:
```diff
@@ -21,7 +21,9 @@ extension TaggedEnumSwitcherVariable {
     ///   - coder: The decoder for cases.
     ///   - context: The context in which to perform the macro expansion.
-    ///   - default: Whether default case is needed.
+    ///   - default: Whether default case is needed. Note that for Bool type,
```
> Can you make the changes to fix above scenario?

@soumyamahunt Absolutely, though don’t hold your breath for lightning-fast updates from me 🥺

qizh replied:
@soumyamahunt, 

- [x] I've fixed the [`TaggedEnumSwitcherVariable.swift` implementation](Sources/PluginCore/Variables/Enum/Switcher/TaggedEnumSwitcherVariable.swift)
- [x] And I've added the new [`expansionWithPartialBoolCoverage`](https://github.com/SwiftyLab/MetaCodable/commit/a405f1b6459db95013eb0eedf39a6709979f88c0#diff-d1738430cb391a485d818e34b05bfdafb847ec0a070492a626898b01471aabc8R20) test in the new [`CodedAsMixedTypesTests.swift`](Tests/MetaCodableTests/CodedAs/CodedAsMixedTypesTests.swift) for your specific example.

qizh committed:
Update `CHANGELOG.md`

qizh committed:
Fix: Handle partial `Bool` coverage in `switch` statements

- [x] Fix the `Bool` `switch` optimization to only skip the `default` case when BOTH `true` and `false` values are explicitly present in the switch.
  - Previously, the default case was incorrectly skipped for all `Bool` switches, breaking macro expansion when `@CodedAs` includes `Bool` values but not all `enum` cases have `Bool` mappings (partial coverage).
- [x] Add logic to detect both `true` and `false` values before skipping default
- [x] Add `CodedAsMixedTypesTests.swift` test for partial `Bool` coverage scenario with mixed types
- [x] Update inline documentation to clarify the behavior

qizh commented:
@copilot or @codex, can some of you ~~guys~~ 🤖👾 review this?

qizh committed:
test: add coverage tests for HelperCoder and related types

Add tests for previously uncovered code paths:
- HelperCoder: decodeIfPresent, encode, encodeIfPresent
- ConditionalCoder: decodeIfPresent, encodeIfPresent
- DefaultSequenceElementCoding: optional encoding/decoding methods
- DynamicCodableIdentifier: CodingKey conformance methods

Also adds:
- .markdownlint.json configuration based on existing .md file patterns
- Tests/TODO/CoverageTracking.md for tracking test coverage progress

qizh committed:
Merge branch 'merge/upstream'

qizh committed:
test: add @Suite declarations and meaningful test names to all test files

Updated 50 test files to:
- Add @Suite declarations for logical grouping of tests
- Replace unnamed @Test() with @Test("descriptive name")
- Group related tests into nested suites where appropriate
- Improve test organization and readability

All tests continue to pass after the updates.

qizh committed:
test: improve all test names with meaningful, unique descriptions

Updated all 565 tests across 50 files:
- Analyzed actual test logic to generate descriptive names
- Ensured all test names are unique across the entire test suite
- Applied proper capitalization and English grammar
- Preserved parameterized tests with arguments:
- Replaced generic names like "expansion" with context-specific descriptions
  e.g., "Generates @Codable conformance for struct with 'open' access"
- Error tests now clearly state what error they verify
  e.g., "Reports error when @CodingKeys is used without @Codable"

Test names now immediately communicate what each test verifies,
making test failures instantly understandable without needing to
inspect the test code.

Updated Tests/TODO/CoverageTracking.md to document the test
organization improvements.

qizh committed:
test: add comprehensive tagging system for all tests

Added Swift Testing tags to organize and filter the 529 tests:

- Created Tags.swift with 30+ tag definitions covering:
  • Macro tags (@Codable, @CodedAt, @CodedIn, etc.)
  • Test type tags (macroExpansion, encoding, decoding, errorHandling)
  • Feature tags (helperCoders, dynamicCoding, generics, etc.)
  • Type tags (structs, classes, enums, rawRepresentable)
  • Helper coder specific tags
  • Coverage tag for coverage-improvement tests

- Tagged all 529 tests based on actual content analysis
- Tags enable filtering tests by macro, feature, or test type
- Example: @Test("...", .tags(.codable, .macroExpansion, .structs))

Updated Tests/TODO/CoverageTracking.md with tagging documentation
and usage examples.

All 565 tests continue to pass with tags applied.

qizh committed:
style: format long @Test declarations as multiline for readability

Added multiline formatting for @Test declarations with many tags:

- Tests with many tags (>3) now use multiline format
- Each tag on its own line for better readability
- Applied to CodableTests and GenericsTests as examples
- Format:
  @Test(
      "Test name",
      .tags(
          .tag1,
          .tag2,
          .tag3
      )
  )

All 565 tests continue to pass.

qizh committed:
docs: fix markdown lint error - add blank line before code block

qizh committed:
fix: handle partial Bool coverage in TaggedEnumSwitcherVariable

- Add tests for partial Bool coverage in CodedAsMixedTypesTests.swift
- Fix decodeSwitchExpression to only skip default case when both
  true and false Bool values are covered
- Previously, default case was skipped for ALL Bool switches, causing
  unreachable code warnings when only one Bool value was specified

The fix checks if both true and false are present in the enum cases
before deciding to skip the default case in Bool switches.

qizh committed:
fix: handle partial Bool coverage in TaggedEnumSwitcherVariable

- Add tests for partial Bool coverage in CodedAsMixedTypesTests.swift
- Fix decodeSwitchExpression to only skip default case when both
  true and false Bool values are covered
- Previously, default case was skipped for ALL Bool switches, causing
  unreachable code warnings when only one Bool value was specified

The fix checks if both true and false are present in the enum cases
before deciding to skip the default case in Bool switches.

(cherry picked from commit bb8f9a5edb7afbdba1fe06e125100d1df4805ae7)

qizh committed:
fix: remove merge conflict markers from `CodedAsMixedTypesTests.swift`

Remove leftover merge conflict markers (`<<<<<<< HEAD, =======, >>>>>>>`) from the test file that were accidentally committed during the previous merge.

qizh committed:
Merge origin/main into tests/improve/coverage/30-01-26

qizh committed:
Add Swift CI workflow for build and test automation

qizh committed:
Disable main.yml workflow in favor of swift.yml

qizh committed:
Re-enable `main.yml` workflow by removing the `.disabled` extension.

This workflow will be disabled on GitHub.

qizh committed:
chore: simplify Swift workflow name to "Swift `build` & `test`"

qizh committed:
Update `README.md`

With another badge @qizh is interested in

qizh committed:
Fix: Downgrade to Swift 6.1 to avoid compiler crash in macro expansion

Swift 6.2.1 has a bug causing fatalError during ExpandSynthesizedMemberMacroRequest
when processing @Codable macro on protocols (ExtPost in PostPage.swift)

qizh committed:
Fix: Use default Xcode toolchain instead of swift-actions/setup-swift

Swift 6.1/6.2 from swift-actions has compiler crash during macro expansion.
Default Xcode toolchain on macos-latest should be more stable.

qizh committed:
Merge pull request #2 from qizh/tests/improve/coverage/30-01-26

Improve test coverage and organization

qizh committed:
Refactor test organization with nested suites and improve test descriptions

- [x] Reorganize `ConditionalCoderTests`, `DefaultSequenceElementCodingTests`, and `DynamicCodableIdentifierTests` into nested test suites.
- [x] Update test descriptions to be more concise and descriptive.
- [x] Remove `CHANGELOG` entry about unreachable default case warnings (again 🤦🏻‍♂️).
- [x] Simplify multi-line test attribute formatting in `CodableTests`.

qizh committed:
Update `swift-syntax` dependency upper bound to support version `603.x`
